### ISSUE TITLE
Unable to hijack thread in empty while (true) loop

### ISSUE THREAD

**Issue by wesrupert** [dotnet/coreclr#4623]
One of our tests was checking funceval timeout on a function with the body `while (true) { ; }`. This timed out properly on windows but wouldn't terminate on OSX or Ubuntu.

**Comment by mikem8361** [comment]
I'm assuming you are calling ICorDebugEval::Abort or RudeAbort to terminate the funceval?

**Comment by wesrupert** [comment]
Yes and yes. We call `ICorDebugEval::Abort`, then if we don't get a response after a few seconds or get an error we call `ICorDebugEval::RudeAbort`. It looks like we also perform further damage control if we don't get a response from the rude abort after a while and end up hanging on an indefinite wait on an async break request, something else that I've noticed doesn't play nice with `while(true) { ; }` statements.

**Comment by mikem8361** [comment]
As it turns out, this wasn't implemented for Linux/OSX. On windows it relies on thread suspend/resume to hijack the busy waiting thread which don't exist on Linux and on OSX the thread suspend/resume isn't exactly compatible with the windows version. Aditya went through hoops to get this scenario to work with GC. Even though some of his work can be leveraged there is still a lot work to do for the func eval abort case.

**Comment by wesrupert** [comment]
Ok. In the meantime, on our end we can add a final fallback to abort the app when this scenario occurs. Otherwise I'm not sure what more we are able to do. Is this a feature that you are interested in investing in before RTM?

**Comment by mikem8361** [comment]
I don’t know yet if we will able to do this before RTM.

**Comment by mikem8361** [comment]
@tzwlai FYI.  This is a new feature/workitem.

**Comment by tzwlai** [comment]
I don't think we can fit this in before RTM.  I'll move this out to 1.1 unless there is strong objection.

**Comment by mikem8361** [comment]
@wesrupert did a customer actually hit this problem or is it just a test failure.

**Comment by wesrupert** [comment]
We have not yet had any customers report hitting this issue. Unsurprisingly, not many people use empty `while (true)` loops in production.

**Comment by chaami** [comment]
The issue happens in tight loops as well....

**Comment by webczat** [comment]
I am not sure how all the things work, but on at least linux you should be able to stop a thread at any moment. I think dotnet debugger does not have any use for ptrace... or it maybe does?

**Comment by mikem8361** [comment]
For .NET Core apps, the managed code needs to be a "safe" point to suspend/debug. A "safe" point in the JIT'ed code is locations in the managed function that the runtime can suspend allowing the managed state to be examined.  

The managed debugging doesn't currently use ptrace.  The only place ptrace might be used someday is mixed mode debugging on Linux where both managed and native code can be debugged in the same process.

**Comment by mikem8361** [comment]
Tight loops may not have any safe places.

**Comment by jkotas** [comment]
> Tight loops may not have any safe places.

The JIT is expected to guarantee that all loops have GC safe points in them. Otherwise, the GC would hang if there is any such loop running in the system.

**Comment by webczat** [comment]
I just meant that if you use some way to suspend threads on windows, there are ways on linux at least too, for example, sending a sigstop signal to thread.
